### PR TITLE
generate CDO directly by binding `libcdo_driver`

### PR DESCRIPTION
**Pull request by makslevental**
This PR implements a "direct" CDO generation utility - ie direct bindings and calls to `libcdo_driver.so` (instead of emitting C++, compiling+linking against `libcdo_driver.so`, and then running).

It works by building `libxaienginecdo` _and statically linking it_. The point of statically linking is to minimize the number of floating libs that need to be in the environment in order for things to work. In addition `libcdo_driver.so` itself (which `libxaienginecdo` depends on) is "re-exported" to `${CMAKE_INSTALL_PREFIX}/lib` using cmake's `install(IMPORTED_RUNTIME_ARTIFACTS ...)`. As far as I can there are no other (non-ubiquitous) runtime dependencies. 

The relevant source file is [lib/Targets/AIETargetCDODirect.cpp](https://github.com/Xilinx/mlir-aie/blob/link_lib_cdo/lib/Targets/AIETargetCDODirect.cpp)

Currently there's only a smoke test (through python bindings) but I plan to basically dump all currently generated bins (generating using the other way) and diff against what's producing with this path.

EDIT: `cmp` tests are checked in.

cc @MaheshRavishankar

### CODE REVIEW

File: test/python/direct_cdo.py
Comment (fifield) [inline test/python/direct_cdo.py:1]: needs copyright/license header

**Review by fifield — APPROVED** [review]
I don't think we need both `aie-generate-cdo` and `aie-generate-cdo-direct`. If the direct cdo gen is working we can replace the old method.

File: test/Targets/AIEGenerateCDODirect/01_precompiled_core_function.mlir [hunk 1]
```diff
@@ -0,0 +1,34 @@
+// (c) Copyright 2023 Advanced Micro Devices, Inc.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// REQUIRES: cdo_direct_generation
+//
+// RUN: export BASENAME=$(basename %s)
+// RUN: rm -rf *.elf* *.xclbin *.bin $BASENAME.cdo_direct $BASENAME.prj
+// RUN: %python aiecc.py --aie-generate-cdo --no-compile-host --tmpdir $BASENAME.prj %s
+// RUN: mkdir $BASENAME.cdo_direct && cp $BASENAME.prj/*.elf $BASENAME.cdo_direct
```
Comment (stephenneuendorffer) [inline test/Targets/AIEGenerateCDODirect/01_precompiled_core_function.mlir:9]: If/when we rip out aiecc.py --aie-generate-cdo, how do we test these?  Can we independently run these tests and (additionally) compare with the --aie-generate-cdo flow?

File: include/aie/Targets/cdo_driver.h [hunk 2]
```diff
@@ -0,0 +1,23 @@
+// 67d7842dbbe25473c3c32b93c0da8047785f30d78e8a024de1b57352245f9689
+/* cdo_driver.h  */
+#ifndef _CDO_DRIVER_H_
+#define _CDO_DRIVER_H_
```
[the remark this answers is not in the document]
Comment (makslevental) [reply include/aie/Targets/cdo_driver.h:4]: This file does exist at e.g. `/Xilinx/Vitis/2023.2/aietools/include/cdo_driver.h`. I can thread it through.

File: lib/Targets/AIETargetCDODirect.cpp [hunk 3]
```diff
@@ -0,0 +1,679 @@
...
+}
+
+#define DEBUG_TYPE "aie-generate-cdo-direct"
+
+using namespace mlir;
+using namespace xilinx;
+using namespace xilinx::AIE;
+
+#define AIERC_STR(x) x, #x
+static const std::map<AieRC, std::string> AIERCTOSTR = {
```
[the remark this answers is not in the document]
Comment (makslevental) [reply lib/Targets/AIETargetCDODirect.cpp:60]: mapping to what and how? aie-rt is strictly a C API library isn't it? So the enum is there and there's fairly good logging around private APIs but no there's no mapping from enum variant to something like `char *`.

File: lib/Targets/AIETargetCDODirect.cpp [hunk 4]
```diff
@@ -0,0 +1,679 @@
...
+    TRY_XAIE_API_FATAL_ERROR(XAie_SetIOBackend, &devInst, XAIE_IO_BACKEND_CDO);
+    TRY_XAIE_API_FATAL_ERROR(XAie_UpdateNpiAddr, &devInst, npiAddr);
+  }
+
+  LogicalResult addErrorHandlingToCDO() {
+    TRY_XAIE_API_LOGICAL_RESULT(XAie_ErrorHandlingInit, &devInst);
+    return success();
+  }
+
+  LogicalResult addAieElfToCDO(uint8_t col, uint8_t row,
```
[the remark this answers is not in the document]
Comment (makslevental) [reply lib/Targets/AIETargetCDODirect.cpp:210]: the aie-rt has things like `XAie`. I'm sticking with that for consistency.

File: test/Targets/AIEGenerateCDODirect/01_precompiled_core_function.mlir (hunk 1 above)
Comment (makslevental) [reply test/Targets/AIEGenerateCDODirect/01_precompiled_core_function.mlir:9]: As jeff says, we don't need both and so we won't even need these tests. This battery of tests is already like comparing a xerox-copy to an already xeros-copied copy of the canonical thing (i.e., I've found one or two bugs in the current CDO generation in this process).

File: lib/Targets/AIETargetCDODirect.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (stephenneuendorffer) [reply lib/Targets/AIETargetCDODirect.cpp:60]: Should the code actually live there though?  https://github.com/stephenneuendorffer/aie-rt/tree/phoenix_v2023.2

[the remark this answers is not in the document]
Comment (makslevental) [reply lib/Targets/AIETargetCDODirect.cpp:60]: It's a `std::map` - how can it live there? Regardless, it's just a map from enum variant to string i.e., not the largest amount of technical debt I've ever taken on...

**Pull request merged**